libsel4/tools/generator_c.py

```diff
@@ -0,0 +1,584 @@
+        return result
+
+    def _gen_unmarshal_result(self, num_mrs, output_params, structs) -> str:
+        result = []
+        source_words = {}
+        for i in range(MAX_MESSAGE_LENGTH):
+            if i < num_mrs:
+                source_words["w%d" % i] = "mr%d" % i
+            else:
+                source_words["w%d" % i] = "seL4_GetMR(%d)" % i
+        unmashalled_params = self.generate_unmarshal_expressions(output_params)
+        for (param, words) in unmashalled_params:
+            if param.type.pass_by_reference():
+                members = struct_members(param.type, structs)
+                for i in range(len(words)):
+                    result.append("\t%s->%s = %s;" %
+                                  (param.name, members[i], words[i] % source_words))
+            else:
+                if param.type.double_word:
+                    result.append("\tresult.%s = ((%s)%s + ((%s)%s << 32));" %
+                                  (param.name, Generator_C.TYPES[64], words[0] % source_words,
+                                   Generator_C.TYPES[64], words[1] % source_words))
+                else:
+                    for word in words:
+                        result.append("\tresult.%s = %s;" % (param.name, word % source_words))
+        # XXX return is generated in _gen_return_end_function
+        # result.append("\treturn result;")
+        # result.append("}")
+        result = "\n".join(result)
+        self.contents.append(result)
+        return result
+
+    def generate_unmarshal_expressions(self, params):
+        wordsize = self.wordsize
+
+        def unmarshal_single_param(first_bit, num_bits, wordsize):
+            """
+            Unmarshal a single parameter.
+            """
+            first_word = first_bit // wordsize
+            bit_offset = first_bit % wordsize
+
+            # Multiword type?
+            if num_bits > wordsize:
+                result = []
+                for x in range(num_bits // wordsize):
+                    result.append("%%(w%d)s" % (x + first_word))
+                return result
+
+            # Otherwise, bit packed.
+            if num_bits == wordsize:
+                return ["%%(w%d)s" % first_word]
+            elif bit_offset == 0:
+                return ["(%%(w%d)s & %#x)" % (
+                    first_word, (1 << num_bits) - 1)]
+            else:
+                return ["(%%(w%d)s >> %d) & %#x" % (
+                    first_word, bit_offset, (1 << num_bits) - 1)]
+
+        # Get their marshalling positions
+        positions = get_parameter_positions(params, wordsize)
+
+        # Generate the unmarshal code.
+        results = []
+        for (param, first_bit, num_bits) in positions:
+            results.append((param, unmarshal_single_param(first_bit, num_bits, wordsize)))
+        return results
+
+    def generate_marshal_expressions(self, params, num_mrs, structs):
+        wordsize = self.wordsize
+        WORD_CONST_SUFFIX_BITS = {
```

Comment:
`<kernel>include/basic_types.h` has the macro `SEL4_WORD_CONST(x)` to hide this. I wonder if we can simply use this here and keep the generator agnostic of this.

Reply:
This seems to be problematic. We've got `basic_types.h` implemented here:
https://github.com/seL4/seL4/blob/master/include/basic_types.h

`libsel4` doesn't include anything from the global kernel directory, but rather things from inside `libsel4` directory, and the paths for includes can be viewed here:
https://github.com/seL4/seL4/blob/master/libsel4/CMakeLists.txt#L71-L78

I expect that this is by design, and placing any include from `../` wouldn't be welcome inside `libsel4/CMakeLists.txt`
However, just to test if things would work, I added the missing path, and then path to every file that was reported in errors.
I ended up having these:
```
    "${CMAKE_CURRENT_SOURCE_DIR}/../include"
    "${CMAKE_CURRENT_SOURCE_DIR}/../include/${KernelWordSize}"
    "${CMAKE_CURRENT_SOURCE_DIR}/../include/arch/${KernelArch}"
    "${CMAKE_CURRENT_SOURCE_DIR}/../include/arch/${KernelArch}/arch/${KernelWordSize}"
```
It won't compile due to some unknown type names and conflicting types.

I think we would benefit from having the macro available from inside `libsel4`, but I expect that the solution would actually require changes in the header files, rather than including files from `../`, and bit a bit more complex than just an addition to this PR.

Reply:
Yes, this needs to be improved in the headers. I will create a PR for this.